packages/form-state-manager/src/utils/manager-api.ts

```diff
@@ -230,6 +232,54 @@ const createManagerApi: CreateManagerApi = ({ onSubmit, clearOnUnmount, initiali
     state.registeredFields.forEach(resetFieldState);
   }
 
+  function validateForm(validate: FormValidator) {
+    const result = formLevelValidator(validate, state.values, managerApi);
+    const currentInvalidFields = Object.keys(state.errors);
+    if (isPromise(result)) {
+      const asyncResult = result as Promise<FormLevelError>;
+      return asyncResult
+        .then(() => {
+          if (!state.validating) {
+            state.errors = {};
+            state.valid = true;
+            state.invalid = false;
+            state.error = undefined;
+            revalidateFields(currentInvalidFields);
+          }
+        })
+        .catch((errors) => {
+          state.errors = errors;
+          state.valid = false;
+          state.invalid = true;
+        });
+    }
+
+    const syncError = result as FormLevelError | undefined;
+    if (syncError) {
+      Object.keys(syncError).forEach((name) => {
+        handleFieldError(name, false, syncError[name]);
+      });
+      state.errors = syncError;
+      state.valid = false;
+      state.invalid = true;
+    } else {
+      state.errors = {};
+      state.valid = true;
+      state.invalid = false;
+      state.error = undefined;
+      /**
+       * Fields have to be revalidated on field level to synchronize the form and field errors
+       */
+      revalidateFields(currentInvalidFields);
```

Comment:
Can we avoid this? This is not how the form currently behaves (check it [here](https://stackblitz.com/edit/jeh3gr?file=index.js)) and it brings performance issues (why to validate fields that have been already validated?)

Also, it means that active field is being validated 2x: field validation > form validation > field validation

Reply:
It definitely can and should be. I wanted to postpone this after we have the memorize functionality implemented because that is a big part of it.

The validation will have have to run twice at some point (it does in the example as well) because the validation functions are different.

Once we have the memo we can "save" the active field revalidation by running the field level after the form level and check if it was validated in that cycle.